### DISCUSSION TITLE
Prisma stops sending @default for @id field after adding a relation

### DISCUSSION THREAD

**Discussion by colincoller** [prisma/prisma#25095] [Q&A]
### Question

I'm writing a simple multi-tenant app and I've noticed what seems like a bug - or at least an extremely confusing behavior - where Prisma stops providing a default value if I add a relation.

My app has an `Account` model representing the account (tenant) and a `User` model representing the account's users.

Each user belongs to exactly 1 account, so there's a FK from `User.accountId` to `Account.id`.

Each account has 0..1 "root user", so there _should be_ a FK from `Account.rootUserId` to `User.id`. 

With just the first FK (from `User.accountId` to `Account.id`), the schema looks like this:
```prisma
model Account {
  id String @id @default(uuid())

  users User[]

  rootUserId String? // <---- This should have a FK to User.id
}

model User {
  id String @default(uuid())

  account   Account @relation(fields: [accountId], references: [id], onDelete: Cascade)
  accountId String

  emailAddress String

  @@id([accountId, id])
  @@unique(name: "accountId_emailAddress", [accountId, emailAddress])
}
```

This generates the SQL:
```
CREATE TABLE "Account" (
    "id" TEXT NOT NULL PRIMARY KEY,
    "rootUserId" TEXT
);

CREATE TABLE "User" (
    "id" TEXT NOT NULL,
    "accountId" TEXT NOT NULL,
    "emailAddress" TEXT NOT NULL,

    PRIMARY KEY ("accountId", "id"),
    CONSTRAINT "User_accountId_fkey" FOREIGN KEY ("accountId") REFERENCES "Account" ("id") ON DELETE CASCADE ON UPDATE CASCADE
);

CREATE UNIQUE INDEX "User_accountId_emailAddress_key" ON "User"("accountId", "emailAddress");
```

With this schema, I can successfully create new accounts:
```js
  const account = await prisma.account.create({
    data: {},
  });
  console.log(account);

  // prisma:query INSERT INTO `main`.`Account` (`id`) VALUES (?) RETURNING `id` AS `id`, `rootUserId` AS `rootUserId`
  //
  // { id: 'f111f614-e39f-4819-8b86-fca0566802cf', rootUserId: null }
```

Note that Prisma supplies the default value for `Account.id` in the `INSERT` above. `Account.id` doesn't have a default constraint in the database. This is fine, things are working. ✅ 

Now, to add the second FK (from `Account.rootUser` to `User.id`), I change the schema like so:
```prisma
model Account {
  id String @id @default(uuid())

  users User[]

  rootUser   User?   @relation("RootUserFK", fields: [id, rootUserId], references: [accountId, id]) // <---- NEW
  rootUserId String?

  @@unique([id, rootUserId]) // <---- NEW
}

model User {
  id String @default(uuid())

  account   Account @relation(fields: [accountId], references: [id], onDelete: Cascade)
  accountId String

  rootAccount Account? @relation("RootUserFK") // <---- NEW

  emailAddress String

  @@id([accountId, id])
  @@unique(name: "accountId_emailAddress", [accountId, emailAddress])
}
```

This generates the SQL:
```
CREATE TABLE "Account" (
    "id" TEXT NOT NULL PRIMARY KEY,
    "rootUserId" TEXT,
    CONSTRAINT "Account_id_rootUserId_fkey" FOREIGN KEY ("id", "rootUserId") REFERENCES "User" ("accountId", "id") ON DELETE RESTRICT ON UPDATE CASCADE
);

CREATE TABLE "User" (
    "id" TEXT NOT NULL,
    "accountId" TEXT NOT NULL,
    "emailAddress" TEXT NOT NULL,

    PRIMARY KEY ("accountId", "id"),
    CONSTRAINT "User_accountId_fkey" FOREIGN KEY ("accountId") REFERENCES "Account" ("id") ON DELETE CASCADE ON UPDATE CASCADE
);

CREATE UNIQUE INDEX "Account_id_rootUserId_key" ON "Account"("id", "rootUserId");

CREATE UNIQUE INDEX "User_accountId_emailAddress_key" ON "User"("accountId", "emailAddress");
```

The schema validates and the SQL looks good. However, when I run the same code, I get an error:
```
prisma:query INSERT INTO `main`.`Account` DEFAULT VALUES RETURNING `id` AS `id`, `rootUserId` AS `rootUserId`
PrismaClientKnownRequestError:
Invalid `prisma.account.create()` invocation in
repro/script.ts:11:40

→ 11 const account = await prisma.account.create(
Null constraint violation on the fields: (`id`)
    at Ln.handleRequestError (repro/node_modules/@prisma/client/runtime/library.js:121:7753)
    at Ln.handleAndLogRequestError (repro/node_modules/@prisma/client/runtime/library.js:121:7061)
    at Ln.request (repro/node_modules/@prisma/client/runtime/library.js:121:6745)
    at async l (repro/node_modules/@prisma/client/runtime/library.js:130:9633) {
  code: 'P2011',
  clientVersion: '5.19.0',
  meta: { modelName: 'Account', constraint: [ 'id' ] }
}
```

Prisma is no longer supplying the default value for `Account.id` in the `INSERT`!

Before:
```
INSERT INTO `main`.`Account` (`id`) VALUES (?) RETURNING `id` AS `id`, `rootUserId` AS `rootUserId`
```

After:
```
INSERT INTO `main`.`Account` DEFAULT VALUES RETURNING `id` AS `id`, `rootUserId` AS `rootUserId`
```

This is no good. 🛑 

I can't for the life of me figure out why the schema changes caused defaults to break. I can (and will) work around the problem by updating my code to supply an explicit `id` value on insert, computing the uuid myself, but that feels like a hack.

### Expected behavior (optional)

Prisma continues to supply the default value after I've added the second FK.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model Account {
  id String @id @default(uuid())

  users User[]

  rootUser   User?   @relation("RootUserFK", fields: [id, rootUserId], references: [accountId, id]) // <---- NEW
  rootUserId String?

  @@unique([id, rootUserId]) // <---- NEW
}

model User {
  id String @default(uuid())

  account   Account @relation(fields: [accountId], references: [id], onDelete: Cascade)
  accountId String

  rootAccount Account? @relation("RootUserFK") // <---- NEW

  emailAddress String

  @@id([accountId, id])
  @@unique(name: "accountId_emailAddress", [accountId, emailAddress])
}
```

```ts
  const account = await prisma.account.create({
    data: {},
  });
```

- OS: macOS
- Database: SQLIte
- Node.js version: 20.16.0
- Prisma version:
    ```
    prisma                  : 5.19.0
    @prisma/client          : 5.19.0
    Computed binaryTarget   : darwin-arm64
    Operating System        : darwin
    Architecture            : arm64
    Node.js                 : v20.16.0
    Query Engine (Node-API) : libquery-engine 5fe21811a6ba0b952a3bc71400666511fe3b902f (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
    Schema Engine           : schema-engine-cli 5fe21811a6ba0b952a3bc71400666511fe3b902f (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
    Schema Wasm             : @prisma/prisma-schema-wasm 5.19.0-31.5fe21811a6ba0b952a3bc71400666511fe3b902f
    Default Engines Hash    : 5fe21811a6ba0b952a3bc71400666511fe3b902f
    Studio                  : 0.502.0
    ```

**Comment by colincoller** [comment]
Running this through ChatGPT suggests
> When you add a foreign key relation in Prisma, the foreign key constraint requires explicit values for related fields. Prisma assumes that if you're managing relations, you likely intend to control the ID values yourself, rather than relying on automatically generated defaults. As a result, it stops automatically sending the @default value for the @id field, expecting that the developer will now manage the assignment of these values manually to maintain referential integrity across the tables involved.

and

> The specific behavior where Prisma stops sending the default value when a relation is added isn't explicitly documented in the Prisma documentation. However, it is implied by the way Prisma handles relational mappings and referential integrity. When you add a relation, Prisma expects you to manage the values of the related fields explicitly to ensure that the foreign key constraints are respected. This expectation can lead to situations where default values are no longer automatically applied as they would be in a non-relational context.

I'm willing to believe that's the case, and so my workaround of generating the uuid manually feels less like a workaround, but if that _is_ the case then the tooling should warn that the `@default` is going to be ignored. 🤔

**Comment by ludralph** [comment]
Hi @colincoller 👋 

This is an excerpt from the [documentation](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#relational-databases-2:~:text=dbgenerated(...)%20function%20when%20you%20use%20introspection.-,Default%20values%20are%20not%20allowed%20on%20relation%20fields%20in%20the%20Prisma,a%20relation%20will%20mean%20that%20relation%20is%20populated%20automatically%20for%20you.,-Default%20values%20can%20be%20used%20with%20scalar):

> Default values are not allowed on relation fields in the Prisma schema. Note however that you can still define default values on the fields backing a relation (the ones listed in the fields argument in the @relation attribute). A default value on the field backing a relation will mean that relation is populated automatically for you.

This is working as intended.

**Reply by redgeparamify** [reply]
I have a similar problem and want to know how this documentation supports that this is working as expected. It seems to me that the @default was added to a "field backing a relation" not the relation field itself. To me this would support the fact that it should work as expected, where the default should be called since the field is simply one referenced in the fields argument of the relation. Am I missing something here?
